arrows/klv/klv_read_write.txx

```diff
@@ -298,6 +361,110 @@ klv_ber_oid_length( T value )
   return ( _int_bit_length( value ) + 6 ) / 7;
 }
 
+// ---------------------------------------------------------------------------
+template < class Iterator >
+double
+klv_read_imap( double minimum, double maximum, Iterator& data, size_t length )
+{
+  KLV_ASSERT_UINT8_ITERATOR( data );
+
+  // Section 8.1.2
+  _check_range_length( minimum, maximum, length );
+
+  auto const int_value = klv_read_int< uint64_t >( data, length );
+  auto value = static_cast< double >( int_value );
+
+  // Section 8.2.2
+  // Left-shift required to shift a bit from the least significant place to the
+  // most significant place
+  auto const msb_shift = length * 8 - 1;
+
+  // Most significant bit and any other bit set means this is a special value
+  if( int_value & ( 1ull << msb_shift ) && int_value != ( 1ull << msb_shift ) )
+  {
+    // Third most significant bit = sign bit
+    bool const sign_bit = int_value & 1ull << ( msb_shift - 2 );
+
+    // Second, fourth, and fifth most significant bits = special value
+    // identifiers
+    constexpr uint64_t identifier_mask = 0xB; // 01011
+    auto const identifier =
+      ( int_value >> ( length * 8 - 5 ) ) & identifier_mask;
+
+    switch( identifier )
+    {
+      case 0x9: // 1001
+        value = std::numeric_limits< double >::infinity();
+        break;
+      case 0xA: // 1010
+        value = std::numeric_limits< double >::quiet_NaN();
+        break;
+      case 0xB: // 1011
+        value = std::numeric_limits< double >::signaling_NaN();
+        break;
+      default: // Reserved and user-defined values
+        return std::numeric_limits< double >::quiet_NaN();
+    }
+
+    return sign_bit ? -value : value;
+  }
+
+  // Normal value
+  auto const terms = _calculate_imap_terms( minimum, maximum, length );
+  value = terms.backward_scale * ( value - terms.zero_offset ) + minimum;
+
+  // Return exactly zero if applicable, overriding rounding errors. IMAP
+  // specification considers this important
+  auto const precision = klv_imap_precision( minimum, maximum, length );
+  return ( std::abs( value ) < precision / 2.0 ) ? 0.0 : value;
+}
+
+// ---------------------------------------------------------------------------
+template < class Iterator >
+void
+klv_write_imap( double value, double minimum, double maximum, Iterator& data,
+                size_t length )
+{
+  KLV_ASSERT_UINT8_ITERATOR( data );
+
+  // Section 8.1.2
+  _check_range_length( minimum, maximum, length );
+
+  // Out-of-range values set to infinity
+  if( value < minimum )
+  {
+    value = -std::numeric_limits< double >::infinity();
+  }
+  else if( value > maximum )
+  {
+    value = std::numeric_limits< double >::infinity();
+  }
+
+  // Section 8.2.1
+  uint64_t int_value = 0;
+  if( std::isnan( value ) )
```

Comment:
It would be marginally more efficient to check this *first* (before checking for `value` not in [`minimum`, `maximum`]).

I'm not sure if it makes more sense to check for literal infinity first or not; hopefully the compiler can optimize that. OTOH, given the limits are finite, you could also just write (pseudocode):
```
if value < minimum
  return negative_infinity
[else] if value > maximum
  return positive_infinity
```

...and dispense with the `isfinite` check altogether.